# Information about data used for nQuire analysis

## Running nQuire
* Sujan ran nQuire while processing and generating the .bam files 
* He generated results using depths of 20, 35, and 50
* He ran nQuire on the full genome and just the CDS
* He did not retain the number of SNPs used for each sample

## Data files
* nQuire results
  * at NERSC
    * `/global/homes/g/grabowsp/data/switchgrass/nquire_8X/Ploidy_338Samples_nquire.txt`
  * at HA
    * `/home/t4c1/WORK/grabowsk/data/switchgrass/nquire_8X/Ploidy_338Samples_nquire.txt`
  * Uploaded 10/22/2020
* Info about library sequencing coverage
  * at NERSC:
    * `/global/homes/g/grabowsp/data/switchgrass/nquire_8X/Pvirg_1070G_reads_mapped.txt`
  * at HA:
    * `/home/t4c1/WORK/grabowsk/data/switchgrass/nquire_8X/Pvirg_1070G_reads_mapped.txt`

## Scratch space
* on NERSCE:
  * `/global/cscratch1/sd/grabowsp/sg_8X_scratch/nquire_scratch`

## Consolidated nQuire results file
* at NERSC
  * `/global/homes/g/grabowsp/data/switchgrass/nquire_8X/consolidated_nQuire_results.txt`
