# Instructions and descriptions of steps for generating genotype files
#  for different analyses of 4X/8X samples

## Background
* Raw genotypes that I will most often use are tetrasomic genotypes for all 
samples generated by Sujan
  * Good for 8X, not correct for 4X
  * 4X genotypes are adjusted when `genlight` objects are generated
* Use CDS SNPs because they seem to be highest quality based on previous 
experience
  * There seem to be fewer mapping issues with CDS SNPs 

## Steps
* Generate chromosome CDS VCFs for sample set
  * Split into 100k-line subfiles in same process
* Generate header file to be used with subfiles
* Generate `genlight` object for each chromosome
* Make genome-wide, subsampled `genlight` object

## Files and Resource Used
### Needed Resources
#### conda environments
* `gen_bioinformatics` environment
  * includes VCFtools which is needed for filtering VCFs
  * `/global/homes/g/grabowsp/.conda/envs/gen_bioinformatics`
* `r_adegenet_env` environment
  * includeds R and adegenet package
  * `/global/homes/g/grabowsp/.conda/envs/r_adegenet_env`
#### Scripts from github
* `make_Chr_genlight_objs.r`
  * used for making chromosome 'genlight' genotype objects
  * `/PATH/TO/REPOSITORY/sg_8X/adegenet_analysis/adegenet_genotype_generation/make_Chr_genlight_objs.r`
* `get_tot_nSNPs.r`
  * Used to quickly get the number of SNPs in each chromosome genlight object
 * `/PATH/TO/REPOSITORY/sg_8X/adegenet_analysis/adegenet_genotype_generation/get_tot_nSNPs.r`
* `subsample_genlight.r`
  * used for making subsampled genome-wide 'genlight' object
  * `/PATH/TO/REPOSITORY/sg_8X/adegenet_analysis/adegenet_genotype_generation/subsample_genlight.r`
#### R-Function files from github
* `general_functions.r`
  * contains functions used in many types of analyses
  * `/PATH/TO/REPOSITORY/sg_8X/general_r_tools/general_functions.r`
* `adegenet_functions.r`
  * contains functions specific to analyses using adegenet package
  * `/PATH/TO/REPOSITORY/sg_8X/adegenet_analysis/adegenet_genotype_generation/adegenet_functions.r`

### Needed Files
#### Starting-point VCFs
* Tetrasomic SNP VCFs for all libraries for each chromosome
  * Found in this directory:
    * `/global/cscratch1/sd/grabowsp/sg_8X_scratch/orig_sujan_files/tetrasomic_vcfs`
  * Transferred over from HA, where originals are stored
#### BED file of CDS location
* CDS BED
  * `/global/homes/g/grabowsp/data/switchgrass/sg_v5_CDS.bed`
#### Library names to include
* Make a file with the library names of the samples to be included
#### Library names for different ploidies
* Tetraploid libraries
  * `/global/homes/g/grabowsp/data/switchgrass/metadata_8X/tet_samps_Nov2020.txt`
* Octoploid libraries
  * `/global/homes/g/grabowsp/data/switchgrass/metadata_8X/oct_samps_Nov2020.txt`
## Generate chromosome CDS VCFs and split into 100k line subfiles
* I generally use maximum 20% missing data (--max-missing 0.8)
* Requires file of library names to keep in a single column
* Easiest way to to submit jobs to cluster
### Submit job
```
cd /PATH/TO/DATA/BEING/USED

sbatch generate_Chrset_sampleset_vcf.sh
```
### Example of submit script
```
#!/bin/bash
#SBATCH -D .
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=4
#SBATCH -A plant
#SBATCH -C haswell
#SBATCH --mem=32G
#SBATCH --qos=genepool_shared
#SBATCH -t 48:00:00
#SBATCH --mail-user pgrabowski@hudsonalpha.org
#SBATCH --mail-type=BEGIN
#SBATCH --mail-type=FAIL
#SBATCH --mail-type=END

module load python/3.7-anaconda-2019.07
source activate /global/homes/g/grabowsp/.conda/envs/gen_bioinformatics

# Directory with base VCFs
DATA_DIR=/global/cscratch1/sd/grabowsp/sg_8X_scratch/orig_sujan_files/tetrasomic_vcfs/

# BED file of CDS SNP intervals
BED_FILE=/global/homes/g/grabowsp/data/switchgrass/sg_v5_CDS.bed

# File of Library Names to keep
SAMP_FILE=/global/homes/g/grabowsp/data/switchgrass/metadata_8X/all_samp_names.txt

# Directory where new VCFs will be saved
OUT_DIR=/global/cscratch1/sd/grabowsp/sg_8X_scratch/allsamps_tet_vcfs/

# sample set name to be used in fame files
SAMPSET=allsamps

cd $OUT_DIR

for CHROM in 01K 01N 02K 02N 03K 03N 04K 04N 05K 05N;
  do
  vcftools --gzvcf \
    $DATA_DIR'Pvirgatum_1070g_Chr'$CHROM'.5alleles.snp.sort.norepeats.vcf.gz' \
    --max-missing 0.8
    --stdout --bed $BED_FILE --keep $SAMP_FILE --recode --recode-INFO-all | \
    gzip -c > $OUT_DIR'Chr'$CHROM'.tetrasomic.CDS.'$SAMPSET'.vcf.gz';
  gunzip -kc $OUT_DIR'Chr'$CHROM'.tetrasomic.CDS.'$SAMPSET'.vcf.gz' | \
    split -l 100000 -d - $OUT_DIR'Chr'$CHROM'.tetrasomic.CDS.'$SAMPSET'.vcf_';  
  done
```

## Generate sample header file to be used with subfiles
* 100k-line VCF subfiles require header when loading them into R for analysis
* 1 header file can work for all the sub-files
  * VCF header information begins with # so isn't loaded into R, so can use
same approach for all subfiles, even those at beginning of chromsome that
have the standard VCF header
* Use the very first sub-file from any chromosome to make the generic header
file
  * Example: `Chr01K...vcf_00`
* The number of lines to be used with `head` depends on the amount of
information in the header section of the VCF
  * For current formatting, there are only 5 lines before the genotypes begin
### Example of generating file
```
# go to directory with genotype files
cd /global/cscratch1/sd/grabowsp/sg_8X_scratch/allsamps_tet_vcfs

SAMPSET=allsamps

gunzip -kc Chr01K.tetrasomic.CDS.$SAMPSET'.vcf.gz' | head -634 | \
tail -1 > CDS.tetrasomic.$SAMPSET'.vcf.sampheader.txt'
```

## Generate `genlight` object for each chromosome
* `Genlight` object is genotypes formated for `adegenet` R package
* Uses `make_Chr_genlight_objs.r` script to generate `genlight` object for
each chromosome
  * hardcoded to include 2 files that have the library names of 8X and 4X samples
    * `/global/homes/g/grabowsp/data/switchgrass/metadata_8X/tet_samps_Nov2020.txt`
    * `/global/homes/g/grabowsp/data/switchgrass/metadata_8X/oct_samps_Nov2020.txt`
* Have to decide what MAF cutoff to use
  * I typically aim for a MAF cutoff where there are ~5 copies of the allele
at the MAF, but using higher MAF is certainly defensible
### Example of submitting job
```
# go to directory with genotypes and submit shell script
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/geo_samps

sbatch gen_geo_genlight.sh
```
### Example of submit script
* From `gen_geo_genlight.sh`
```
#!/bin/bash
#SBATCH -D .
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=4
#SBATCH -A plant
#SBATCH -C haswell
#SBATCH --mem=32G
#SBATCH --qos=genepool_shared
#SBATCH -t 48:00:00
#SBATCH --mail-user pgrabowski@hudsonalpha.org
#SBATCH --mail-type=BEGIN
#SBATCH --mail-type=FAIL
#SBATCH --mail-type=END

module load python/3.7-anaconda-2019.07
source activate /global/homes/g/grabowsp/.conda/envs/adegenet_2_env

DATA_DIR=/global/cscratch1/sd/grabowsp/sg_8X_scratch/geobig_tet_vcfs/

cd $DATA_DIR

HEADER_FILE=/global/cscratch1/sd/grabowsp/sg_8X_scratch/geobig_tet_vcfs/CDS.tetrasomic.geobig.vcf.sampheader.txt

SAMPSET=geobig

MAF_CUT=0.0018

for CHR_N in {01..09};
  do
  for CHR_T in K N;
    do
    TEST_CHR=$CHR_N$CHR_T
    SEARCH_STRING=Chr$TEST_CHR'.tetrasomic.CDS.'$SAMPSET'.vcf_'
    Rscript /global/homes/g/grabowsp/tools/sg_8X/adegenet_analysis/adegenet_genotype_generation/make_Chr_genlight_objs.r \
    $DATA_DIR $SEARCH_STRING'*' $HEADER_FILE $MAF_CUT;
    done;
  done;
```

## Generate genome-wide subsampled `genlight` object
* Subsample SNPs from all chromosomes to get desired number of genome-wide SNPs
### Extract number of SNPs in each chromosome genlight object
* uses R script to make data frame containing file names and SNPs in each file
  * `/global/homes/g/grabowsp/tools/sg_8X/adegenet_analysis/adegenet_genotype_generation/get_tot_nSNPs.r`
#### Example
```
# load module with adegenet
module load python/3.7-anaconda-2019.07
source activate /global/homes/g/grabowsp/.conda/envs/adegenet_2_env

# Set directory where genlight files are stored
DATA_DIR=/global/cscratch1/sd/grabowsp/sg_8X_scratch/geobig_tet_vcfs

# string used for searching for genlight files
FILE_SUB_SHORT=geobig.genlight.rds

# output prefix
OUT_SHORT=geobig

cd $DATA_DIR

Rscript /global/homes/g/grabowsp/tools/sg_8X/adegenet_analysis/adegenet_genotype_generation/get_tot_nSNPs.r \
$DATA_DIR '*'$FILE_SUB_SHORT $OUT_SHORT
```
### Choose subsampling level
* use file generate above to quickly calculate calculate subsampling level.
  * should chose level slightly higher so can then re-sample to exact SNP number
#### Example
* in R
```
# module load python/3.7-anaconda-2019.07
# source activate /global/homes/g/grabowsp/.conda/envs/adegenet_2_env

library(data.table)

# file with number of SNPs in each genlight file
res_file <- '/global/cscratch1/sd/grabowsp/sg_8X_scratch/geobig_tet_vcfs/geobig.SNPcount.txt'
res <- fread(res_file)

# goal number of SNPs
goal_n <- 5e4

goal_n / sum(res$nSNPs)
#[1] 0.005608668
# this is the subsampling level
```

### Generate object
* uses `subsample_genlight.r` script
* Takes a few minutes, but can be run in interactive session
#### Example
```
# load conda environment
module load python/3.7-anaconda-2019.07
source activate /global/homes/g/grabowsp/.conda/envs/adegenet_2_env

# go to directory where want to run script
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/geo_samps

# directory with chromosome genlight objects
DATA_DIR=/global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/geo_samps/

# String used for searching for genlight files
FILE_SUB=geobig.genlight.rds

# name of output file
OUT_SHORT=combo.sub.polyploid.CDS.$SAMPSET'.genlight.rds'

# percentage of total SNPs to select
PER_SUBSAMP=0.07

# Exact number of SNPs to have in final file
TOT_SNP=5e4
 
cd $DATA_DIR

Rscript /global/homes/g/grabowsp/tools/sg_8X/adegenet_analysis/adegenet_genotype_generation/subsample_genlight.r \
$DATA_DIR '*'$FILE_SUB_SHORT $OUT_SHORT $PER_SUBSAMP $TOT_SNP
```


